mdp/nodes/sfa_nodes.py

```diff
@@ -145,22 +208,33 @@ def _stop_training(self, debug=False):
 
         rng = self._set_range()
 
+        ####################################################################### EIG
+
         #### solve the generalized eigenvalue problem
         # the eigenvalues are already ordered in ascending order
-        try:
-            self.d, self.sf = self._symeig(self.dcov_mtx, self.cov_mtx,
-                                     range=rng, overwrite=(not debug))
-            d = self.d
-            # check that we get only *positive* eigenvalues
-            if d.min() < 0:
-                err_msg = ("Got negative eigenvalues: %s."
-                           " You may either set output_dim to be smaller,"
-                           " or prepend the SFANode with a PCANode(reduce=True)"
-                           " or PCANode(svd=True)"% str(d))
-                raise NodeException(err_msg)
-        except SymeigException as exception:
-            errstr = str(exception)+"\n Covariance matrices may be singular."
-            raise NodeException(errstr)
+        if not self._use_svd:
```

Comment:
The right solution here (as mentioned by email) is to do something like
```
if self._use_svd:
    self._symeig = sfa_svd_solver
```

I would to that in the `__init__`, and define `sfa_svd_solver` as a function above `SFANode`.

(My preferred solution is to have a `sfa_solver` parameter in the constructor instead of `use_svd`, so that in principle you can have multiple solver. The default `sfa_solver` would be the `sfa_symeig_solver`. That would avoid the proliferation of on/off parameters if one day you add another solver.)

Reply:
Regarding my preferred solution, to avoid misunderstanding: as far as I'm concerned you can keep the flag, I'm not active anymore and do not want to make design decisions for the new masters of mdp. 

For the other changes requested, I really recommend to follow them as they will spare you some headaches later.

Reply:
I find the `self._symeig = sfa_svd_solver` hackish, because `_symeig` sounds like the field was intended to hold a raw symeig solver and just allow to insert a non default one.
The idea with the `sfa_solver` parameter sounds much better. Consider there is an sfa_solver that uses _symeig internally. Than it would hurt if the sfa node would perform `self._symeig = sfa_solver`, wouldn't it?

Reply:
Maybe we chose the wrong name, that is easy to change. But the attribute and its signature were designed exactly for this case: have a hook to use a different solver. It would be easy to rewrite the `svd_solver` so that it has the same signature:
```
svd_solver(self.dcov_mtx, self.cov_mtx, d=self.d, range=rng, overwrite=True/False)
```
That would make `svd_solver` a standalone function like `symeig`, and we can pass it in the constructor. Nothing else in `_stop_training` needs to change.

Reply:
I guess that's the problem with undocumented parameters - one needs to guess their purpose.

> Maybe we chose the wrong name, that is easy to change

I'm afraid it's not easy to change, at least not if you/we care about breaking exiting code. We don't know who is accessing _symeig (e.g. Varun pointed out on the mailing list that he is using it.). It's a private field, so we can actually rename it - people must be prepared for this, right? Still, it wouldn't be so nice. I'd suggest to do the renaming in this manner:
Have `_sfa_solver` as a replacement for `_symeig`, while keeping `_symeig`. `_symeig` continues as being a switch for the raw symeig solving component (I suppose there are plenty eig and symeig routines, so this isn't so bad to have, e.g. there are GPU based solvers etc).
Then by default we init `self._sfa_solver = self._symeig` and use `_sfa_solver` in `_stop_training`.
This would introduce precise naming while not breaking existing code that uses `_symeig`.

Reply:
One more thing:
`_sfa_solver` sould have the same signature like `_symeig` I suppose, but can we slightly rename the `range` parameter? It is currently hiding the range function.

Reply:
I would remove the `self._symeig` member and replace it with something like `self._eig_solve` which would be a parameter of the `__init__()` constructor, default to `eig_solve='symeig', and accept `'SVS'` and/or any other solutions on offer (e.g. Stefan's alternative).

To me this should be a nice compromise between your suggestions.

Reply:
But why break backwards compatibility just for the sake of consistency? Just leave the name `_symeig` alone and replace its value with whatever solver you like. Regarding the `range` parameter, well, again, I don't think it's worth it to break backwards compatibility. The name comes from the parameter name in the LAPACK routine. But I don't see the problem with the `range` function. The name is overriden only in the function scope, and even there you still have  `__builtins__.range` if you really need it, no?

Reply:
Right, backwards compatibility should definitely not be broken. In that case using `_symeig` to hold / point to whatever solver is chosen seems a good solution.

Just for the sake of argument though, how would renaming `_symeig` break backwards compatibility.. it's a private member that shouldn't be accessed from the outside either way?

Reply:
Well, we have seen that Varun was using the private attribute already, no? So, people were using it even if it was a private attribute. Maybe we should have made it public from the beginning, I don't know. Nothing stops us from creating a new public property `eig_solver` or whatever, and then automatically set the `_symeig` attribute when `eig_solver` is changed, so that you get a nicer and public interface for the future, without breaking old code?

Reply:
Hm. If we get this (or @Stewori's) PR through before Varun's, then he'll rebase it anyways and it won't matter what we call it. Of course someone else might have used the `_symeig` from the outside but.. well, they shouldn't have, as it's marked as a private attribute. Maybe someone else can pitch in on this, if there's a consensus to preserve private attributes like this then I'd be okay with that.